### PR TITLE
Add feature to pull interceptors from Ballerina Central

### LINKED ISSUE #901: Creating a Policy Hub for the Microgateway for the purpose of sharing policies
### Describe your problem(s)

In order to provide the API developers from anywhere to share their policies with
others and reuse policies contributed by others, a facility will be needed.

### Describe your solution

Creating a facility to pull interceptors from the Ballerina Central.

### How will you implement it

Interceptors are stored inside a Ballerina Module and the aforesaid module is being published inside the Ballerina Central. The developers are able to provide the module name along with it's respective organization name and lastly the interceptor function which will be referred from that module. In addition the developers are able to define the version of the module, if the interceptors are being referred from the module versions apart from the latest version. The suggested OpenAPI definition's format is as follows.

If the interceptors are to be referred from a specific version of the module:
- x-wso2-request-interceptor: <organization_name>/<module_name>/<version_of_module>:<interceptor_function_name>
- x-wso2-response-interceptor: <organization_name>/<module_name>/<version_of_module>:<interceptor_function_name>

 If the interceptors are to be referred from the latest version of the module:
- x-wso2-request-interceptor: <organization_name>/<module_name>:<interceptor_function_name>
- x-wso2-response-interceptor: <organization_name>/<module_name>:<interceptor_function_name>

During the microgateway project build time, the module mentioned in the swagger definition will be pulled from the Ballerina Central.
---

### PR DESCRIPTION
**Pull request by SabinaPrasadhi**
### Purpose

The objective is to create a developer community around the Microgateway. As the API developers use policies for performing various types of transformations, validations and so on at the API gateway, this includes a facility to provide API developers from anywhere to share their policies with others and reuse policies contributed by others as well.

### Issues

 #901

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

OS: Linux
JDK version: openjdk version "1.8.0_222"

---
#### Maintainers: Check before merge
- [ ] Assigned 'Type' label
- [ ] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java [hunk 1]
```diff
@@ -77,6 +81,7 @@ public BallerinaOperation buildContext(Operation operation, ExtendedAPI api) thr
             return getDefaultValue();
         }
 
+        //PrintStream outStream = System.out;
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]: Shall we remove commented code lines?

Comment (SabinaPrasadhi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]: Noted. Will remove them.

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java [hunk 2]
```diff
@@ -253,6 +324,7 @@ public void setApiRequestInterceptor(String requestInterceptor) {
         // api level interceptor
         if (this.requestInterceptor == null || !this.requestInterceptor.equals(requestInterceptor)) {
             this.apiRequestInterceptor = requestInterceptor;
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java [hunk 3]
```diff
@@ -274,4 +346,6 @@ public void setBasicAuth(BasicAuth basicAuth) {
             this.basicAuth = basicAuth;
         }
     }
+
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java [hunk 4]
```diff
@@ -274,4 +346,6 @@ public void setBasicAuth(BasicAuth basicAuth) {
             this.basicAuth = basicAuth;
         }
     }
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 5]
```diff
@@ -599,23 +599,37 @@ private static void validateInterceptorAvailability(String interceptorName, bool
...
-        }
-    }
+     private static void validateAPIRequestInterceptors(OpenAPI openAPI, String openAPIFilePath) {
+         if (openAPI.getExtensions() != null) {
+             Optional<Object> apiRequestInterceptor = Optional
+                     .ofNullable(openAPI.getExtensions().get(OpenAPIConstants.REQUEST_INTERCEPTOR));
+             apiRequestInterceptor.ifPresent(
+                     requestInterceptorValue -> validateInterceptorAvailability(requestInterceptorValue.toString(),
+                             true, openAPIFilePath, null, null));
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 6]
```diff
@@ -650,29 +664,63 @@ private static void validateSingleResourceExtensions(Operation operation, String
...
+        }
+
+        // validate response interceptor for a single resource
+        // skip validation if the response interceptor is being looked-up from the Ballerina Central
+        if (!responseInterceptor.toString().contains(OpenAPIConstants.BALLERINA_CENTRAL_KEYWORD)) {
+            validateSingleResourceResponseInterceptors(operation, pathItem, operationName, openAPIFilePath);
+        }
     }
 
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 7]
```diff
@@ -650,29 +664,63 @@ private static void validateSingleResourceExtensions(Operation operation, String
...
      * @param pathItem        path name
      * @param operationName   operation name
      * @param openAPIFilePath file path to openAPI definition
      */
-    private static void validateSingleResourceInterceptors(Operation operation, String pathItem, String operationName,
-                                                           String openAPIFilePath) {
-        //validate request interceptor
+    private static void validateSingleResourceRequestInterceptors(Operation operation, String pathItem,
+                                                                  String operationName, String openAPIFilePath) {
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 8]
```diff
@@ -650,29 +664,63 @@ private static void validateSingleResourceExtensions(Operation operation, String
...
         Optional<Object> requestInterceptor = Optional.ofNullable(operation.getExtensions()
                 .get(OpenAPIConstants.REQUEST_INTERCEPTOR));
-        requestInterceptor.ifPresent(value -> validateInterceptorAvailability(value.toString(), true,
-                openAPIFilePath, pathItem, operationName));
-        //validate response interceptor
+
+        requestInterceptor.ifPresent(
+                requestInterceptorValue -> validateInterceptorAvailability(requestInterceptorValue.toString(),
+                        true, openAPIFilePath, pathItem, operationName));
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 9]
```diff
@@ -827,10 +875,27 @@ private static void validateAPINameAndVersion(OpenAPI openAPI, String openAPIFil
      * @param openAPIFilePath file path to openAPI definition
      */
     public static void validateOpenAPIDefinition(OpenAPI openAPI, String openAPIFilePath, String openAPIVersion) {
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java [hunk 10]
```diff
@@ -650,29 +664,63 @@ private static void validateSingleResourceExtensions(Operation operation, String
...
+     * Validate Resource Response interceptors for Single Resource
+     *
+     * @param operation       {@link Operation} object
+     * @param pathItem        path name
+     * @param operationName   operation name
+     * @param openAPIFilePath file path to openAPI definition
+     */
+    private static void validateSingleResourceResponseInterceptors(Operation operation, String pathItem,
+                                                                   String operationName, String openAPIFilePath) {
+
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java]:
```suggestion
```

File: components/micro-gateway-core/.factorypath [hunk 11]
```diff
@@ -0,0 +1,157 @@
+<factorypath>
```
Comment (praminda) [inline components/micro-gateway-core/.factorypath]: Remove this file.

File: tests/.factorypath [hunk 12]
```diff
@@ -0,0 +1,220 @@
+<factorypath>
```
Comment (praminda) [inline tests/.factorypath]: Remove this file

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java [hunk 13]
```diff
@@ -157,18 +234,40 @@ private void setPaths(OpenAPI openAPI) throws BallerinaServiceGenException {
...
                         .generateBasicAuthFromSecurity(this.api.getMgwApiSecurity()));
+
+                //set the import modules specified in the operation level request interceptors
+                setImportModules(operation.getValue().getRequestInterceptorModule());
+
+                //set the import modules specified in the operation level response interceptors
+                setImportModules(operation.getValue().getResponseInterceptorModule());
+
                 //if it is the developer first approach
+
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java]:
```suggestion

```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java [hunk 14]
```diff
@@ -192,12 +291,14 @@ private void setPaths(OpenAPI openAPI) throws BallerinaServiceGenException {
...
                     //to set scope property of API
                     operation.getValue().setScope(api.getMgwApiScope());
+
                 }
             });
             paths.add(new AbstractMap.SimpleEntry<>(path.getKey(), balPath));
         }
     }
 
+
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java]:
```suggestion

```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java [hunk 15]
```diff
@@ -192,12 +291,14 @@ private void setPaths(OpenAPI openAPI) throws BallerinaServiceGenException {
                     });
                     //to set scope property of API
                     operation.getValue().setScope(api.getMgwApiScope());
+
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java]:
```suggestion

```

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java [hunk 16]
```diff
@@ -129,6 +136,76 @@ public Info getInfo() {
...
+        ArrayList<String> requestInterceptorStatement = new ArrayList<>();
+        ArrayList<String> requestInterceptorModule = new ArrayList<>();
+
+        Optional<Object> apiRequestInterceptor = Optional.ofNullable(openAPI.getExtensions()
+                .get(OpenAPIConstants.REQUEST_INTERCEPTOR));
+
+        if (apiRequestInterceptor.toString().contains(OpenAPIConstants.BALLERINA_CENTRAL_KEYWORD)) {
+            requestInterceptorStatement.add(apiRequestInterceptor.toString());
+            requestInterceptorStatement.forEach(statement -> {
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java]: `requestInterceptorStatement` always going to contain only one item. So do we need `forEach`. Or do we need `requestInterceptorStatement` and `requestInterceptorModule` to be a list in the first place?

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java [hunk 17]
```diff
@@ -129,6 +136,76 @@ public Info getInfo() {
...
+
+        ArrayList<String> responseInterceptorStatement = new ArrayList<>();
+        ArrayList<String> responseInterceptorModule = new ArrayList<>();
+
+        Optional<Object> apiResponseInterceptor = Optional.ofNullable(openAPI.getExtensions()
+                .get(OpenAPIConstants.RESPONSE_INTERCEPTOR));
+
+        if (apiResponseInterceptor.toString().contains(OpenAPIConstants.BALLERINA_CENTRAL_KEYWORD)) {
+            responseInterceptorStatement.add(apiResponseInterceptor.toString());
+            responseInterceptorStatement.forEach(statement -> {
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaService.java]: Same as above

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/OpenAPIConstants.java [hunk 18]
```diff
@@ -15,4 +17,14 @@
     public static final String THROTTLING_TIER = "x-wso2-throttling-tier";
     public static final String DISABLE_SECURITY = "x-wso2-disable-security";
     public static final String AUTHORIZATION_HEADER = "x-wso2-auth-header";
+    public static final String INTERCEPTOR_STATEMENT_SEPARATOR = ":";
+    public static final String MODULE_STATEMENT_SEPARATOR = "/";
+    public static final String WHITESPACE = " ";
```
Comment (praminda) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/OpenAPIConstants.java]: I don't think we need a constant for this. We can use `' '` where necessary.

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/OpenAPIConstants.java [hunk 19]
```diff
@@ -15,4 +17,14 @@
     public static final String THROTTLING_TIER = "x-wso2-throttling-tier";
     public static final String DISABLE_SECURITY = "x-wso2-disable-security";
     public static final String AUTHORIZATION_HEADER = "x-wso2-auth-header";
+    public static final String INTERCEPTOR_STATEMENT_SEPARATOR = ":";
+    public static final String MODULE_STATEMENT_SEPARATOR = "/";
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/OpenAPIConstants.java]: I don't think we need to define constants for these.

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java [hunk 20]
```diff
@@ -110,14 +114,47 @@ public BallerinaOperation buildContext(Operation operation, ExtendedAPI api) thr
...
-            //set resource level request interceptors
+
+            /*
+            Set resource level request interceptors and set the ballerina module to be imported if specified.
+            */
             Optional<Object> requestInterceptor = Optional.ofNullable(extensions
                     .get(OpenAPIConstants.REQUEST_INTERCEPTOR));
-            requestInterceptor.ifPresent(value -> this.requestInterceptor = value.toString());
-            //set resource level response interceptors
+            if (requestInterceptor.toString().contains(OpenAPIConstants.MODULE_STATEMENT_SEPARATOR)) {
```
Comment (AmaliMatharaarachchi) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/template/service/BallerinaOperation.java]: You can move this to a method and use it for both request and response interceptor.

**Pull request merged**